File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/PhotoDataRepositoryImpl.java
package nl.acidcats.tumblrlikes.data_impl.photodata;

import android.support.annotation.Nullable;
import android.util.Log;

import java.io.File;
import java.util.List;

import javax.inject.Inject;

import nl.acidcats.tumblrlikes.core.constants.FilterType;
import nl.acidcats.tumblrlikes.core.models.Photo;
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository;
import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.PhotoDataGateway;
import nl.acidcats.tumblrlikes.datalib.BuildConfig;
import rx.Observable;
import rx.schedulers.Schedulers;

/**
 * Created by stephan on 11/04/2017.
 */

public class PhotoDataRepositoryImpl implements PhotoDataRepository {
    private static final String TAG = PhotoDataRepositoryImpl.class.getSimpleName();

    private PhotoDataGateway _photoDataGateway;
    private long _startViewTime;
    private long _currentPhotoId;
    private final boolean _debug = BuildConfig.DEBUG;
    private FilterType _currentFilterType = FilterType.UNHIDDEN;

    @Inject
    public PhotoDataRepositoryImpl(PhotoDataGateway photoDataGateway) {
        _photoDataGateway = photoDataGateway;
        setFilterType(_currentFilterType);
    }

    @Override
    public boolean hasPhoto(long postId) {
        return _photoDataGateway.hasPhoto(postId);
    }

    @Override
    public List<Photo> storePhotos(List<Photo> photos) {
        _photoDataGateway.storePhotos(photos);

        return photos;
    }

    @Override
    public List<Photo> getAllPhotos() {
        return _photoDataGateway.getAllPhotos();
    }

    @Override
    public long getPhotoCount() {
        return _photoDataGateway.getPhotoCount();
    }

    @Override
    public Photo getNextPhoto() {
        return _photoDataGateway.getNextPhoto();
    }

    @Override
    public boolean hasUncachedPhotos() {
        return _photoDataGateway.hasUncachedPhotos();
    }

    @Override
    public Photo getNextUncachedPhoto() {
        return _photoDataGateway.getUncachedPhoto();
    }

    @Override
    public void markAsCached(long id, String path) {
        _photoDataGateway.setPhotoCached(id, true, path);
    }

    private Void uncachePhoto(Photo photo) {
        if (_debug) Log.d(TAG, "uncachePhoto: " + photo.getFilePath());

        if (photo.getFilePath() == null) return null;

        //noinspection ConstantConditions
        File file = new File(photo.getFilePath());
        if (file.exists()) {
            if (_debug) Log.d(TAG, "uncachePhoto: file exists");

            if (file.delete()) {
                if (_debug) Log.d(TAG, "uncachePhoto: file deleted");

                _photoDataGateway.setPhotoCached(photo.getId(), false, null);
            }
        }

        return null;
    }

    @Override
    public Observable<Void> removeCachedHiddenPhotos() {
        return Observable
                .just(_photoDataGateway.getCachedHiddenPhotos())
                .subscribeOn(Schedulers.io())
                .flatMapIterable(photoEntities -> photoEntities)
                .map(this::uncachePhoto)
                .toList()
                .flatMap(photoEntities -> Observable.just(null));
    }

    @Override
    public void setPhotoViewStartTime(long id, long currentTime) {
        if (_debug) Log.d(TAG, "startPhotoView: id = " + id);

        _currentPhotoId = id;

        _startViewTime = currentTime;
    }

    @Override
    public void updatePhotoViewTime(long id, long currentTime) {
        if (_debug) Log.d(TAG, "endPhotoView: id = " + id + ", _currentPhotoId = " + _currentPhotoId);

        if (id == 0 || id != _currentPhotoId) {
            if (_debug) Log.d(TAG, "endPhotoView: no photo found to end");
            return;
        }

        _photoDataGateway.addPhotoViewTime(id, currentTime - _startViewTime);
    }

    @Override
    public void setPhotoLiked(long id, boolean isLiked) {
        _photoDataGateway.setPhotoLiked(id, isLiked);
    }

    @Override
    public void setPhotoFavorite(long id, boolean isFavorite) {
        _photoDataGateway.setPhotoFavorite(id, isFavorite);
    }

    @Override
    public void hidePhoto(long id) {
        _photoDataGateway.setPhotoHidden(id);

        Photo photo = _photoDataGateway.getPhotoById(id);
        if (photo != null) {
            uncachePhoto(photo);
        }
    }

    @Override
    @Nullable
    public Photo getPhotoById(long id) {
        return _photoDataGateway.getPhotoById(id);
    }

    @Override
    public void setFilterType(FilterType filterType) {
        _currentFilterType = filterType;

        _photoDataGateway.initFilter(filterType);
    }

    @Override
    public FilterType getFilterType() {
        return _currentFilterType;
    }

    @Override
    public boolean isPhotoCacheMissing(Photo photo) {
        String filePath = photo.getFilePath();

        if (filePath != null) {
            File file = new File(filePath);
            return !file.exists();
        }

        return false;
    }

    @Override
    public List<Long> setPhotosUncached(final List<Long> idList) {
        _photoDataGateway.setPhotosCached(idList, false);

        return idList;
    }

    @Override
    public List<Photo> getCachedPhotos() {
        return _photoDataGateway.getCachedPhotos();
    }
}

File: datalib/src/main/java/nl/acidcats/tumblrlikes/data_impl/photodata/PhotoDataRepositoryImpl.kt
package nl.acidcats.tumblrlikes.data_impl.photodata

import android.util.Log
import nl.acidcats.tumblrlikes.core.constants.FilterType
import nl.acidcats.tumblrlikes.core.models.Photo
import nl.acidcats.tumblrlikes.core.repositories.PhotoDataRepository
import nl.acidcats.tumblrlikes.data_impl.photodata.gateway.PhotoDataGateway
import rx.Observable
import rx.schedulers.Schedulers
import java.io.File
import javax.inject.Inject

/**
 * Created on 25/10/2018.
 */
class PhotoDataRepositoryImpl @Inject constructor(private val photoDataGateway: PhotoDataGateway) : PhotoDataRepository {

    private lateinit var currentFilterType: FilterType
    private var currentPhotoId: Long = 0
    private var startViewTime: Long = 0

    init {
        setFilterType(FilterType.UNHIDDEN)
    }

    override fun hasPhoto(postId: Long): Boolean = photoDataGateway.hasPhoto(postId)

    override fun storePhotos(photos: List<Photo>): List<Photo> {
        photoDataGateway.storePhotos(photos)

        return photos
    }

    override fun getAllPhotos(): List<Photo> = photoDataGateway.allPhotos

    override fun getPhotoCount(): Long = photoDataGateway.photoCount

    override fun getNextPhoto(): Photo? = photoDataGateway.nextPhoto

    override fun setPhotoViewStartTime(id: Long, currentTime: Long) {
//        Log.d(TAG, "setPhotoViewStartTime: ")

        currentPhotoId = id;

        startViewTime = currentTime;
    }

    override fun updatePhotoViewTime(id: Long, currentTime: Long) {
        if (id != 0L && id == currentPhotoId) {
            photoDataGateway.addPhotoViewTime(id, currentTime - startViewTime)
        }
    }

    override fun setPhotoLiked(id: Long, isLiked: Boolean) = photoDataGateway.setPhotoLiked(id, isLiked)

    override fun setPhotoFavorite(id: Long, isFavorite: Boolean) = photoDataGateway.setPhotoFavorite(id, isFavorite)

    override fun hidePhoto(id: Long) {
        photoDataGateway.setPhotoHidden(id)

        val photo = photoDataGateway.getPhotoById(id)
        if (photo != null) {
            uncachePhoto(photo)
        }
    }

    private fun uncachePhoto(photo: Photo): Boolean {
        photo.filePath ?: return false

        val file = File(photo.filePath)
        if (file.exists() && file.delete()) {
            photoDataGateway.setPhotoCached(photo.id, false, null)
        }

        return true
    }

    override fun getPhotoById(id: Long): Photo? = photoDataGateway.getPhotoById(id)

    override fun setFilterType(filterType: FilterType) {
        currentFilterType = filterType

        photoDataGateway.initFilter(filterType)
    }

    override fun getFilterType(): FilterType = currentFilterType

    override fun hasUncachedPhotos(): Boolean = photoDataGateway.hasUncachedPhotos()

    override fun getNextUncachedPhoto(): Photo? = photoDataGateway.uncachedPhoto

    override fun markAsCached(id: Long, path: String) = photoDataGateway.setPhotoCached(id, true, path)

    override fun removeCachedHiddenPhotos(): Observable<Boolean> {
        return Observable.fromCallable { photoDataGateway.cachedHiddenPhotos }
                .subscribeOn(Schedulers.io())
                .flatMapIterable { it }
                .map { uncachePhoto(it) }
                .toList()
                .map { true }
    }

    override fun isPhotoCacheMissing(photo: Photo): Boolean {
        val filePath = photo.filePath ?: return false

        return !File(filePath).exists()
    }

    override fun setPhotosUncached(idList: List<Long>): List<Long> {
        photoDataGateway.setPhotosCached(idList, false)

        return idList
    }

    override fun getCachedPhotos(): List<Photo> = photoDataGateway.cachedPhotos
}